### PR TITLE
Pic 497 pss requirements endpoint

### PR DESCRIPTION
**Pull request by jhevans**

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/delius/data/api/PssRequirement.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package uk.gov.justice.digital.delius.data.api;
+
+import lombok.Builder;
+import lombok.Getter;
+
+@Getter
+@Builder
+public class PssRequirement {
+    private Long pssRequirementId;
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/digital/delius/data/api/PssRequirement.java:9]: Maybe start adding some swagger to these

File: src/main/java/uk/gov/justice/digital/delius/service/RequirementService.java [hunk 2]
```diff
@@ -29,18 +37,55 @@ public ConvictionRequirements getRequirementsByConvictionId(String crn, Long con
...
     }
 
-    private Stream<Requirement> getRequirementStreamFromDisposal(Disposal disposal) {
-        return disposal != null && disposal.getRequirements() != null ? disposal.getRequirements().stream() : Stream.empty();
+    public PssRequirements getPssRequirementsByConvictionId(String crn, Long convictionId) {
+        var offender = offenderRepository.findByCrn(crn)
+                .orElseThrow(() -> new NotFoundException(String.format("Offender with CRN '%s' not found", crn)));
+
+        var requirements = getDisposalStream(convictionId, offender)
+                .map(Disposal::getCustody)
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/digital/delius/service/RequirementService.java:53]: very minor, but a bit clunky using a list of Disposals and Custody when there can only be one - but I suppose the result is correct - so perhaps not worry about it

**Review by andymarke — APPROVED** [review]
LGTM

File: src/main/java/uk/gov/justice/digital/delius/service/RequirementService.java (hunk 2 above)
Comment (jhevans) [reply src/main/java/uk/gov/justice/digital/delius/service/RequirementService.java:53]: No this is a good call, it did occur to me that some of this stream processing seems out of place given we're only expecting one element.  I think I started with a list of events and then ran with it but now that you mention it, it probably would make more sense to resolve it to a single event and then only start streaming again when you get the pssRequirements. Will amend 👍

**Review by chris-faulkner-mojdt — APPROVED** [review]
Few comments

File: src/main/java/uk/gov/justice/digital/delius/transformers/RequirementTransformer.java [hunk 3]
```diff
@@ -13,23 +15,56 @@
...
-                .requirementNotes(req.getRequirementNotes())
-                .requirementTypeMainCategory(requirementTypeMainCategoryOf(req.getRequirementTypeMainCategory()))
-                .requirementTypeSubCategory(KeyValueTransformer.keyValueOf(req.getRequirementTypeSubCategory()))
-                .startDate(req.getStartDate())
-                .terminationDate(req.getTerminationDate())
-                .terminationReason(KeyValueTransformer.keyValueOf(req.getTerminationReason()))
-                .length(req.getLength())
-                .lengthUnit(lengthUnitOf(req))
-                .build()).orElse(null);
+        return Optional.ofNullable(requirement)
```
[the remark this answers is not in the document]
Comment (jhevans) [reply src/main/java/uk/gov/justice/digital/delius/transformers/RequirementTransformer.java:18]: It is covered by `RequirementServiceTest` but it's not where you'd expect it which isn't ideal. I didn't want to duplicate effort for the sake of it but I could pare back the RequirementService tests and create dedicated tests for the mapping logic.

Screenshot 2020-09-10 at 09 42 38

[the remark this answers is not in the document]
Comment (chris-faulkner-mojdt) [reply src/main/java/uk/gov/justice/digital/delius/transformers/RequirementTransformer.java:18]: don't want to hold up the PR for this and it was there from before but I  tend to think that all public methods ought to have their own unit tests.

[the remark this answers is not in the document]
Comment (jhevans) [reply src/main/java/uk/gov/justice/digital/delius/transformers/RequirementTransformer.java:18]: To be fair it's a good call and I'm mostly just being lazy 😄  will JFDI

File: src/main/java/uk/gov/justice/digital/delius/transformers/RequirementTransformer.java [hunk 4]
```diff
@@ -13,23 +15,56 @@
...
+                .orElse(null);
+    }
+
+    public static uk.gov.justice.digital.delius.data.api.PssRequirement pssRequirementOf(PssRequirement pssRequirement) {
+        return Optional.ofNullable(pssRequirement)
+                .map(pssr -> uk.gov.justice.digital.delius.data.api.PssRequirement.builder()
+                        .pssRequirementId(pssRequirement.getPssRequirementId())
+                        .type(pssRequirmentTypeMainCategoryOf(pssRequirement.getPssRequirementTypeMainCategory()))
+                        .subType(pssRequirementTypeSubCategoryOf(pssRequirement.getPssRequirementTypeSubCategory()))
+                        .active(pssRequirement.getActiveFlag() == 1L)
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/digital/delius/transformers/RequirementTransformer.java:45]: very minor but you could have used TypesTransformer::zeroOneToBoolean

**Review by andymarke — APPROVED** [review]
LGTM

**Pull request merged**